wnfs/src/private/file.rs

```diff
@@ -921,35 +971,44 @@ impl PrivateForestContent {
     /// Reads a number of bytes starting from a given offset.
     pub async fn read_at<'a>(
         &'a self,
-        offset: usize,
-        len: usize,
+        byte_offset: u64,
+        len_limit: Option<usize>,
         forest: &'a impl PrivateForest,
         store: &'a impl BlockStore,
     ) -> Result<Vec<u8>> {
-        let block_content_size = MAX_BLOCK_CONTENT_SIZE;
-        let chunk_size_upper_bound = (self.get_size_upper_bound() - offset).min(len);
+        let block_content_size = MAX_BLOCK_CONTENT_SIZE as u64;
+        let mut chunk_size_upper_bound = self.get_size_upper_bound() - byte_offset as usize;
+
+        if len_limit.is_some() {
+            chunk_size_upper_bound = chunk_size_upper_bound.min(len_limit.unwrap());
+        }
+
         if chunk_size_upper_bound == 0 {
             return Ok(vec![]);
         }
-        let first_block = offset / block_content_size;
-        let last_block = (offset + len) / block_content_size;
+
+        let first_block = byte_offset / block_content_size;
+        let last_block = len_limit.map(|len| (byte_offset + len as u64) / block_content_size);
+
         let mut bytes = Vec::with_capacity(chunk_size_upper_bound);
         let mut content_stream = Box::pin(self.stream(first_block, forest, store)).enumerate();
+
         while let Some((i, chunk)) = content_stream.next().await {
             let chunk = chunk?;
-            let index = first_block + i;
+            let index = first_block + i as u64;
             let from = if index == first_block {
-                (offset - index * block_content_size).min(chunk.len())
+                (byte_offset - index * block_content_size).min(chunk.len() as u64)
             } else {
                 0
             };
-            let to = if index == last_block {
-                (offset + len - index * block_content_size).min(chunk.len())
+            let to = if Some(index) == last_block {
+                (byte_offset + len_limit.unwrap() as u64 - index * block_content_size)
+                    .min(chunk.len() as u64)
```

Comment:
I see how this `.unwrap()` is valid, but it's not super easy to figure out why. (It's because `Some(index) == last_block` is only true, when `last_block` is `Some(_)`, which is only true if `len_limit` is `Some(_)`.)
```suggestion
            let to = if let Some(len_limit) = len_limit && Some(index) == last_block {
                (byte_offset + len_limit as u64 - index * block_content_size)
                    .min(chunk.len() as u64)
```

Reply:
Yeah makes sense! Unfortunately this gives me an error, says `let` expressions in this position are unstable 👀

Reply:
What version of rustc do you use?

Reply:
Lol nevermind I see it in CI

Reply:
Well, then let's just make it `.unwrap_or_default()` for now:
```rs
            let to = if Some(index) == last_block {
                (byte_offset + len_limit.unwrap_or_default() as u64 - index * block_content_size)
                    .min(chunk.len() as u64)
```

Reply:
![](https://media.giphy.com/media/iDaWdM8gJrjVXd5TmP/giphy.gif)